app/services/foreman/plugin/role_lock.rb

```diff
@@ -0,0 +1,71 @@
+module Foreman
+  class Plugin
+    class RoleLock
+      attr_reader :plugin_id
+
+      def initialize(plugin_id)
+        @plugin_id = plugin_id
+      end
+
+      def register_role(name, permissions, role_registry)
+        Role.transaction do
+          role = process_role name, permissions
+          role.ignore_locking do |r|
+            r.add_permissions!(permissions)
+          end
+          role_registry.role_ids << role.id
+        end
+      end
+
+      def process_role(name, permissions)
+        role = Role.find_by :name => name
+
+        if role && role.permission_diff(permissions).empty?
+          role.update_attribute :origin, @plugin_id unless role.origin
+        elsif role
+          role = rename_and_create role, name, permissions
+        else
+          role = create_plugin_role name, permissions
+        end
+        role
+      end
+
+      def rename_and_create(role, original_name, permissions)
+        rename_existing role, original_name
+        create_plugin_role original_name, permissions
+      end
+
+      def create_plugin_role(name, permissions)
+        role = Role.create :name => name, :origin => @plugin_id
+        role.ignore_locking do |r|
+          r.add_permissions!(permissions)
+        end
+        role
+      end
+
+      def rename_existing(role, original_name)
+        prefix = "Customized"
+        role_name = "#{prefix} #{original_name}"
+        if Role.find_by(:name => role_name)
+          rename_with_free_name role, prefix, original_name
```

Comment:
could this be combined in single method? It seems that this method covers first iteration where num would be nil, so I'd recommend  improving `generate_name` to that num is optional. This also deduplicates line 48 where `generate_name` could be used.